src/serde.rs

```diff
@@ -0,0 +1,184 @@
+use crate::HashBag;
+use core::fmt;
+use core::hash::{BuildHasher, Hash};
+use core::marker::PhantomData;
+use serde::de::{Error, MapAccess, SeqAccess, Visitor};
+use serde::ser::{SerializeSeq, SerializeStruct, Serializer};
+use serde::Deserializer;
+use serde::{Deserialize, Serialize};
+
+pub(crate) struct HashBagVisitor<T, S> {
+    marker: PhantomData<fn() -> HashBag<T, S>>,
+}
+
+impl<T, S> HashBagVisitor<T, S>
+where
+    T: Eq + Hash,
+    S: BuildHasher + Clone,
+{
+    fn new() -> Self {
+        HashBagVisitor {
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<'de, T, S> Visitor<'de> for HashBagVisitor<T, S>
+where
+    T: Deserialize<'de> + Eq + Hash,
+    S: BuildHasher + Clone + Default,
+{
+    type Value = HashBag<T, S>;
+
+    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        formatter.write_str("a HashBag")
+    }
+
+    fn visit_seq<M>(self, mut access: M) -> Result<Self::Value, M::Error>
+    where
+        M: SeqAccess<'de>,
+    {
+        let mut bag: HashBag<T, S> =
+            HashBag::with_capacity_and_hasher(access.size_hint().unwrap_or(0), Default::default());
+
+        while let Some(entry) = access.next_element::<EntryWrapper<T>>()? {
+            bag.insert_many(entry.entry, entry.count);
+        }
+
+        Ok(bag)
+    }
+}
+
+impl<'de, T, S> Deserialize<'de> for HashBag<T, S>
+where
+    T: Deserialize<'de> + Eq + Hash,
+    S: BuildHasher + Clone + Default,
+{
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_seq(HashBagVisitor::<T, S>::new())
+    }
+}
+
+impl<T, H> Serialize for HashBag<T, H>
+where
+    T: Serialize + Eq + Hash,
+    H: BuildHasher + Clone,
+{
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut bag = serializer.serialize_seq(Some(self.len()))?;
+
+        for (entry, count) in self.set_iter() {
+            bag.serialize_element(&EntryWrapper { entry, count })?;
```

Comment:
I think this ends up serializing `&T`, not `T`. Now, that's not really a problem since, I believe, `Serialize` is implemented for `&T` too, but wasn't immediately obvious to me.